[Docs] Further improve the concept-exercises.md document (#1283)
exercism/v3: docs/concept-exercises.md

**Purpose:** Introduce the concept(s) that the exercise teaches to the student.

- The information provided should give the student just enough context to figure out the solution themselves.

> - Don't give any extra information about a concept not needed to understand the fundamentals of the concept or solve the exercise.

Reviewer: I'd like this to be rules about the files, not rules about what people should do. I also think it would be good to have "dos" rather than "do nots". I think the language is less confrontational and more welcome. I'm happy to be disagreed with though.

So, as an example, this would become something more like:

```suggestion
- Only information that is needed to understand the fundamentals of the concept and solve the exercise should be provided. Extra information should be left for the `after.md`.
```

Author: I like it :)

- Links should be used sparingly, if at all. While a link explaining a complex topic like recursion might be useful, for most concepts the links will provide more information than needed so explaining things concisely inline should be the aim.

> - Use the proper technical terms (e.g. a keyword) so that the student can easily search for more information.

Reviewer: ```suggestion
- Proper technical terms should be used so that the student can easily search for more information. For example, (some better example here)...
```

Author: I like it. Now for a better example...

Author: Maybe for now leave out the example?

```suggestion
- Proper technical terms should be used so that the student can easily search for more information.
```

> - Don't use code examples unless the exercise introduces new syntax, in which case an example of the syntax must be included; students should not need to search the web for examples of syntax.

Reviewer: ```suggestion
- Code examples should only be used to introduces new syntax (students should not need to search the web for examples of syntax). In other cases provide descriptions or links instead of code.
```

As an example, the introduction to a "strings" exercise might describe a string as just a "Sequence of Unicode characters" or a "series of bytes", tell the users how to create a string, and explain that a string has methods that can be used to manipulate it. Unless the student needs to understand more nuanced details in order to solve the exercise, this type of brief explanation (along with an example of its syntax) should be sufficient information for the student to solve the exercise.

> See [this example introduction.md file][introduction.md]. Notice how the introduction is very minimal, but that the language-specific keywords are enquoted and an example of the newly introduced syntax is included.

Reviewer 2: "enquoted" is not a very common term IME. Even google suggests "Did you mean _inputted_?". Non-native speakers may struggle with it.

```suggestion
See [this example introduction.md file][introduction.md]. Notice how the introduction is very minimal, but that the language-specific keywords are enclosed in quotes and an example of the newly introduced syntax is included.
```

This file is split into two parts.

1. The first part explains the "story" or "theme" of the exercise. It should generally contain no code samples.
2. The second part provides clear instructions of what a student needs to do to, in the form of one or more tasks.

Each task must conform to the following standard:

> - Start with a third-level heading starting with a number (.e.g `### 1. Do X`, `### 2. Do Y`).

Reviewer: ```suggestion
- Start with a third-level heading starting with a number (e.g. `### 1. Do X`, `### 2. Do Y`).
```
